Add local variables to iocsh

oksanagit opened the pull request:
…Load() and now it is a local macro. #487: new fuctions are added set() and show(), allowing setting and showing local macros

exzombie commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1325,8 +1327,7 @@ iocshCmd (const char *cmd)
 {
-    if (pathname)
-        epicsEnvSet("IOCSH_STARTUP_SCRIPT", pathname);
+// IOCSH_STARTUP_SCRIPT is set inside iocshBody()
```
I don't think this comment is needed. Documenting history of the code in code comments is important when the history affects the present. But this is not the case here, I doubt knowing that `IOCSH_STARTUP_SCRIPT` used to be set here once upon a time will help the future reader.

exzombie commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1367,6 +1368,35 @@ iocshEnvClear(const char *name)
 }
 
+void epicsStdCall
+iocshSet(const char *name, const char * value)
```
Please check for `NULL`, and if `name` is an empty string. Also, consider the return value of `macPutValue` and handle it using `iocshSetError()`.

I think `macPutValue` should check for `NULL` and empty name itself, but I checked and it doesn't. Well, somebody has to :)

exzombie replied:
Speaking of `iocshSetError()`, I see now that this change removes a bunch of error handling calls that were added just yesterday. I believe you were making changes against an older version of the code. Please rebase so that only the desired changes are made part of this pull request.

mdavidsaver replied:
#507 may conflict with #508, #497, and perhaps #459.  @JJL772 fyi.

mdavidsaver replied:
... and yes, also 6ca716a77f24247e96952d1778bb1673600bb420.

exzombie commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1471,7 +1498,30 @@ static const iocshFuncDef iocshLoadFuncDef = {"iocshLoad",2,iocshLoadArgs,
+    iocshLoad(args[0].sval, args[1].sval);
+}
+
+/* set*/
+static const iocshArg iocshSetArg0 = { "pathname",iocshArgString};
+static const iocshArg iocshSetArg1 = { "macros", iocshArgString};
+static const iocshArg *iocshSetArgs[2] = {&iocshSetArg0, &iocshSetArg1};
+static const iocshFuncDef iocshSetFuncDef = {"set",2,iocshSetArgs,
+                                              "Takes local macro name and value\n"
+                                              "Sets local macro with provided value\n"};
+static void iocshSetCallFunc(const iocshArgBuf *args)
+{
+    iocshSet(args[0].sval, args[1].sval);
+}
+
+/* show */
+static const iocshArg iocshShowArg0 = { "pathname",iocshArgString};
+static const iocshArg iocshShowArg1 = { "macros", iocshArgString};
+static const iocshArg *iocshShowArgs[2] = {&iocshShowArg0, &iocshShowArg1};
+static const iocshFuncDef iocshShowFuncDef = {"show",2,iocshShowArgs,
+                                              "Shows all set local marcros\n"};
+static void iocshShowCallFunc(const iocshArgBuf *args)
+{
+    iocshShow();
```
The metadata of these two commands is incomplete. The `show` command takes no arguments, and `iocshShowArgs` should reflect that. The `set` command *does* take two arguments, but the first one is no path, and the second one is not a list of macros. Please update this documentation.

I would avoid refering to "macros" for these user-facing descriptions. From the user's point of view, they are dealing with iocsh variables and don't care that the underlying implementation relies on the macro machinery.

exzombie commented:
@oksanagit, I see that this pull request is marked as a draft, but I chose to review it regardless, hoping to help you finish it. I'm sorry if I have commented on something you already know. But if I was too terse, ask away or stop by :)

anjohnson commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1367,6 +1368,35 @@ iocshEnvClear(const char *name)
     }
 }
 
+void epicsStdCall
+iocshSet(const char *name, const char * value)
```
The new functions `iocshSet()` and `iocshShow()` should be marked `static` and not `epicsStdCall`; they will only be called from within this file. In fact unless they are called from elsewhere in this file (I don't think they are) it would be better to combine their functionality into the `iocshSetCallFunc()` and `iocshShowCallFunc()` routines directly; there's no need for them to even be separate routines.

anjohnson commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1367,6 +1368,35 @@ iocshEnvClear(const char *name)
+        context = (iocshContext *) epicsThreadPrivateGet(iocshContextId);
+
+        if (context != NULL) {
+            macReportMacros(context->handle); // returns long, do we need it?
+            // takes macReportMacros((MAC_HANDLE *handle ), where to take it from?
```
We're going to need to do a bit more for this routine, since the output from `macReportMacros()` is more complex and shows internal information that would only confuse users. That may mean modifying macLib as well, but we can talk about that later.

mdavidsaver replied:
> ... shows internal information that would only confuse users ...

I won't disagree, but does macLib currently provide any other report/dump/show functions?

Can we avoid scope creep into `macCore.c`?

anjohnson replied:
@mdavidsaver Unfortunately no, and looking at the output from `macReportMacros()` using @oksanagit's now-fixed branch we definitely need new code to generate something more like the output from `var` and `epicsEnvShow` instead of this:
```
Cox$ bin/darwin-aarch64/softIoc st.cmd 
# <<< st.cmd ---
echo "IOCSH_STARTUP_SCRIPT = st.cmd"
IOCSH_STARTUP_SCRIPT = st.cmd
show
e name             rawval           value
- ----             ------           -----
s ----             ------           -----
  IOCSH_STARTUP_SCRIPT st.cmd           st.cmd
#
iocshLoad l0.cmd "level=0,TOP_STARTUP_SCRIPT=st.cmd"
# <<< l0.cmd ---
show
e name             rawval           value
- ----             ------           -----
s ----             ------           -----
  IOCSH_STARTUP_SCRIPT st.cmd           st.cmd
s ----             ------           -----
  level            0                0
  TOP_STARTUP_SCRIPT st.cmd           st.cmd
  IOCSH_STARTUP_SCRIPT l0.cmd           l0.cmd
#
iocshLoad l1.cmd "level=1,PARENT=l0.cmd"
# <<< l1.cmd ---
echo "IOCSH_STARTUP_SCRIPT=l1.cmd, level=1"
IOCSH_STARTUP_SCRIPT=l1.cmd, level=1
show
e name             rawval           value
- ----             ------           -----
s ----             ------           -----
  IOCSH_STARTUP_SCRIPT st.cmd           st.cmd
s ----             ------           -----
  level            0                0
  TOP_STARTUP_SCRIPT st.cmd           st.cmd
  IOCSH_STARTUP_SCRIPT l0.cmd           l0.cmd
s ----             ------           -----
  level            1                1
  PARENT           l0.cmd           l0.cmd
  IOCSH_STARTUP_SCRIPT l1.cmd           l1.cmd
# --- l1.cmd >>>
#
show
e name             rawval           value
- ----             ------           -----
s ----             ------           -----
  IOCSH_STARTUP_SCRIPT st.cmd           st.cmd
s ----             ------           -----
  level            0                0
  TOP_STARTUP_SCRIPT st.cmd           st.cmd
  IOCSH_STARTUP_SCRIPT l0.cmd           l0.cmd
# --- l0.cmd >>>
#
echo "IOCSH_STARTUP_SCRIPT = st.cmd"
IOCSH_STARTUP_SCRIPT = st.cmd
show
e name             rawval           value
- ----             ------           -----
s ----             ------           -----
  IOCSH_STARTUP_SCRIPT st.cmd           st.cmd
# --- st.cmd >>>
epics> exit
Cox$ cat st.cmd l0.cmd l1.cmd 
# <<< st.cmd ---
echo "IOCSH_STARTUP_SCRIPT = $(IOCSH_STARTUP_SCRIPT)"
show
#
iocshLoad l0.cmd "level=0,TOP_STARTUP_SCRIPT=$(IOCSH_STARTUP_SCRIPT)"
#
echo "IOCSH_STARTUP_SCRIPT = $(IOCSH_STARTUP_SCRIPT)"
show
# --- st.cmd >>>
# <<< l0.cmd ---
show
#
iocshLoad l1.cmd "level=1,PARENT=$(IOCSH_STARTUP_SCRIPT)"
#
show
# --- l0.cmd >>>
# <<< l1.cmd ---
echo "IOCSH_STARTUP_SCRIPT=$(IOCSH_STARTUP_SCRIPT), level=$(level)"
show
# --- l1.cmd >>>
```
The separator lines labelled `s` in the show output LH column are where each new scope begins. Look at the `show` output from the l1.cmd file; that confuses even me at times.

anjohnson reviewed (changes requested):
I'm out of time for tonight, we can talk more tomorrow.

oksanagit committed:
Address #469 and #487

469: Set IOCSH_STARTUP_SCRIPT in iocshBody(), making it a macLib variable.
487: Add IOCSH commands set and show, to manipulate iocsh macLib variables.

Replaced suggested "show" command with options to the "set" command.

ralphlange commented:
I'm not sure if I understand the commit message
'... IOCSH_STARTUP_SCRIP was moved from iocshLoad() to iocshLoad() ...'

oksanagit commented:
> I'm not sure if I understand the commit message
> '... IOCSH_STARTUP_SCRIP was moved from iocshLoad() to iocshLoad() ...'

This line should be ....IOCSH_STARTUP_SCRIP was moved from iocshLoad() to iocshBody() ... My bad

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1100,6 +1100,8 @@ iocshBody (const char *pathname, const char *commandLine, const char *macros)
 
     macPushScope(handle);
     macInstallMacros(handle, defines);
+     if (pathname)
```
Somehow the indentation is 5 spaces on this line.  Please change to 4.  (the next line has 8, which is correct)

oksanagit commented:
@mdavidsaver , could you point me to the example which might help me to write a unit test for the new set command? I was looking at iocshTest.cpp different tests.  And so far, with try and fail, I had 100% the later. You seem to be the author of this file. thx

oksanagit committed:
Add unit tests

oksanagit commented:
I hope all these failures are because new unit tests were added.

mdavidsaver commented:
I think the tests you want to preform would be something like the following.  Create two cmd files to test scoping, and using environment variables to communicate results out to iocshTest.c.

```c
testFile("iocshTestLocal1.cmd");
testOk1(getenv("innerA") && strcmp(getenv("innerA"), "A outer")==0);
testOk1(getenv("innerB") && strcmp(getenv("innerB"), "B inner")==0);
testOk1(getenv("outerB") && strcmp(getenv("outerB"), "B outer")==0);
```

In `iocshTestLocal1.cmd`

```
set "A" "A outer"
set "B" "B outer"
iocshLoad "iocshTestLocal2.cmd"
epicsEnvSet "outerB" "$(B)"
```

In `iocshTestLocal2.cmd`

```
set "B" "B inner"
epicsEnvSet "innerA" "$(A)"
epicsEnvSet "innerB" "$(B)"
```

oksanagit commented:
When the show() low level support is done I can write unit tests later in another PR @anjohnson

anjohnson commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1474,6 +1476,43 @@ static void iocshLoadCallFunc(const iocshArgBuf *args)
+        context = (iocshContext *) epicsThreadPrivateGet(iocshContextId);
+
+        if (context != NULL) {
+            macPutValue(context->handle, args[0].sval, args[1].sval);
```
I discovered I can segfault the IOC by running the command `set` with no parameters, please add a check for `args[0].sval == NULL`. However it occurred to me that we could omit the `show` command completely, and just have `set` with no parameters display the list of variables, like we already do with `var`.

However, the following commands _should_ be allowed as the second `set` deletes the variable that was created by the first one.
```
set abcd 1234
set abcd
```

anjohnson replied:
I lied about the segfault above, I just found my notes and it required these:
```
set abcd 1234
set
```
The reason is still the same, `macPutValue()` doesn't like the parameter name to be NULL.

oksanagit replied:
@anjohnson I am working on it, and will test. Will write here if have questions.

anjohnson reviewed (commented):
Hi @oksanagit, I worked on the macLib changes at LGA while waiting for my delayed flight on Friday and I have something that works, but it needs some additional modifications to untangle it from the `epicsEnvSet()` routines, and I have to add tests for the new functionality there.

If you add the check for `args[0].sval == NULL` that I discuss below and remove the `show` command completely we could merge this PR without waiting for my additions. I can adjust the NULL test and make that call the code in a later PR.

anjohnson committed:
Add macIterateMacros() API

oksanagit commented:
@anjohnson please take a look, if this is what you meant having at the end. Also, your comment about crashing ioc, was helpful.

anjohnson commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -1474,6 +1476,33 @@ static void iocshLoadCallFunc(const iocshArgBuf *args)
+            if(args[0].sval !=0){
+                macPutValue(context->handle, args[0].sval, args[1].sval);
+            }
+        else {
+            macReportMacros(context->handle);
+        }
```
This `else` block needs to be indented by 4 more spaces.

However I don't think we should expose the output from `macReportMacros()` to our users, I'd rather you just do something like this for now:
```C
    else {
        printf("    'set' with no parameters is still being implemented\n");
    }
```

anjohnson reviewed (commented):
This looks good, just one more small change please, see below.

Thanks!

mdavidsaver commented:
Core Group: In trying to understand this PR and #551, why was the setting of `IOCSH_STARTUP_SCRIPT` moved from `iocshLoad()` to one of the calling functions `iocsh()`?

Is it worth trying to split out this part of the change as a separate PR?  eg. by amending #551

anjohnson committed:
Create osdEnvSet() and osdEnvUnset()

Allows iocsh to set env var's without calling iocshEnvClear().
Move iocshEnvClear() calls to the new epicsEnvSet() wrappers.

The iocsh() routine needs to set IOCSH_STARTUP_SCRIPT without
affecting macro var's.

anjohnson commented:
The description in issue #469 answers the Core Group's question, this was the first bug @oksanagit was fixing — the first commit moved setting `IOCSH_STARTUP_SCRIPT` as a macLib macro into `iocshBody()` within the `macPushScope()` so at the end of an `ioshLoad`ed script (i.e. an `iocshLoad` command inside the main `st.cmd` file) the previous value of it and any other variables set in the command will be restored.

We realized later that we do have to call `epicsEnvSet()` at least once, because the startup script's name needs to remain in the environment after it ends; that explains the third commit, setting it as an environment variable in the `iocsh()` function.

I'm still working (off and on) on an iterate method for macLib...

anjohnson commented:
Release Notes for this updated version of the PR, not committed to avoid conflicts:

### New `set` and `show` commands for iocsh variables

Since EPICS 3.15.1 the IOC shell command `iocshLoad` has been available to run
an iocsh script and pass it `variable=value` settings for iocsh variables that
are limited in scope to that script. However there was no way to set variables
inside a script that would also disappear when it ends.

This release includes the new iocsh commands `set name value` to set an iocsh
variable in the current scope, and `show` to display all iocsh variables visible
in the current scope. Omitting `value` from the `set` command marks that name as
deleted in the current scope without affecting the parent scope.

Scripts that set variables which must be visible after the script ends should
continue to use `epicsEnvSet` or `epicsEnvUnset`, which both delete any iocsh
variables with the same name in all iocsh scopes.

Any iocsh variables given the same name as an environment variable will hide the
environment variable when iocsh expands `${name}` or `$(name)` in the current
and any child scopes; the original value will reappear when the iocsh variable
goes out of scope.

A new macLib routine was added to libCom to support the `show` command above,
`long macIterateMacros(handle, callback, user)` calls `callback()` for each
macro defined in `handle` passing it `user` followed by the macro's name and
fully-expanded value as strings. See macLib.h for details.

anjohnson commented on modules/libcom/src/macLib/macCore.c:
```diff
@@ -465,14 +462,75 @@
+
+        /* There may be new values for this name in later scopes.
+         * Find the latest matching entry visible in any scope. */
+        final = lookup( handle, entry->name, FALSE );
```
The `lookup()` routine will return `entry` if no other macro is defined with the same name in later scopes. The above call should never be able to return `NULL`.

anjohnson commented:
@mdavidsaver I moved the `osdEnvSet()` and `osdEnvUnset()` declarations into iocsh.cpp since that's the only file which has calls to them, and removed all their decorations.

mdavidsaver commented:
> ... moved the `osdEnvSet()` and `osdEnvUnset()` declarations into iocsh.cpp ...

Can you move these into a private header included by both iocsh.cpp and the osdEnv.c files?

Otherwise the compiler can not detect if the function signatures are inconsistent.

How about `module/libcom/src/osi/osdEnvPvt.h`?

anjohnson commented:
Converting to draft for now, I now think this needs some tests for the macIterateMacros() API and maybe combining the new `show` command into `set` (which I will do).

anjohnson committed:
Document the "set" command
